LINUX/mvpp2_netmap.h

```diff
@@ -211,49 +233,45 @@ static int mvpp2_netmap_rxsync(struct netmap_kring *kring, int flags)
 
 		nm_i = kring->nr_hwtail;
 		while (rx_received && nm_i != hwtail_lim) {
+			struct netmap_slot *slot = &ring->slot[nm_i];
 			struct mvpp2_rx_desc *rx_desc = mvpp2_rxq_next_desc_get(rxq);
 			u32 status = mvpp2_rxdesc_status_get(port, rx_desc);
 			int pool = (status & MVPP2_RXD_BM_POOL_ID_MASK) >> MVPP2_RXD_BM_POOL_ID_OFFS;
-			phys_addr_t phys_addr = mvpp2_rxdesc_cookie_get(port, rx_desc);
-			dma_addr_t dma_addr = mvpp2_rxdesc_dma_addr_get(port, rx_desc);
-			void *data = (void *)phys_to_virt(phys_addr);
-			struct page *page = virt_to_page(data);
-			void *addr = NMB(na, &ring->slot[nm_i]);
-			uint64_t offset = nm_get_offset(kring, &ring->slot[nm_i]);
-			int length;
-
-			if (status & (MVPP2_RXD_BUF_HDR|MVPP2_RXD_ERR_SUMMARY))
-			{
+			int length = mvpp2_rxdesc_size_get(port, rx_desc) - MVPP2_MH_SIZE;
+			uint32_t buf_index = (uint32_t)mvpp2_rxdesc_cookie_get(port, rx_desc);
+			phys_addr_t paddr;
+
+			if (status & (MVPP2_RXD_BUF_HDR|MVPP2_RXD_ERR_SUMMARY)) {
 				nm_prerr("NETMAP[%s:%d] rxsync error(0x%08x)\n", ifp->name, rxq->id, status);
 				ifp->stats.rx_errors++;
 				mvpp2_rx_error(port, rx_desc);
 				if (status & MVPP2_RXD_BUF_HDR)
 					mvpp2_buff_hdr_pool_put(port, rx_desc, pool, status);
 				else
-					mvpp2_bm_pool_put(port, pool, dma_addr, phys_addr);
+					mvpp2_bm_pool_put(port, pool, mvpp2_rxdesc_dma_addr_get(port, rx_desc), buf_index);
 				rx_received--;
 				rx_complete++;
 				continue;
 			}
 
-			prefetch(page);
-			length = mvpp2_rxdesc_size_get(port, rx_desc);
-			length -= MVPP2_MH_SIZE;
+			/* Put the received netmap buffer into the slot */
+			BUG_ON(buf_index >= na->na_lut.objtotal);
+			paddr = na->na_lut.plut[buf_index].paddr + MVPP2_MH_SIZE + NETMAP_SLOT_HEADROOM;
+			netmap_sync_map_cpu(na, NULL, (bus_dmamap_t)&paddr, length, NR_RX);
+			uint32_t old_index = slot->buf_idx;
+			slot->buf_idx = buf_index;
+			slot->ll_ofs = MVPP2_MH_SIZE + NETMAP_SLOT_HEADROOM;
+			slot->len = length;
+			slot->flags = 0;
+
+			/* Put the old buffer into the buffer pool */
```

Comment:
Shouldnt something similar be in the lower loop (where it goes over slots userspace is ready to free), wouldnt that enable us to have the maximum possible number of buffers available all the time, rather than waiting for something to overwrite the slot before we free the buffer back to the pool?

Reply:
We already have the maximum amount of buffers available as we allocate extra buffers for the h/w buffer pool. So even if userspace holds on to one of the buffers in the ring we still have 2048 in the percpu buffer pool.

Reply:
Ah okay cool sounds good.